### ISSUE TITLE
Fix for CSS Rendering Issue at large row counts

### ISSUE THREAD

**Issue by pbower** [6pac/SlickGrid#1044]
### Describe the bug

Hi @ghiscoding ,

Hope you're doing well.

Wondering if you have a moment to take a quick look at a bug / suggested fix?

**Problem:**
I recently encountered a CSS rendering issue at large counts:

![image](https://github.com/user-attachments/assets/d03b13d1-4de0-4dd4-b4cd-90c6c6bb03b8)

 See correct styles here for comparison:
  
![image](https://github.com/user-attachments/assets/e17e0ab3-ef48-4967-a77d-d3fe979c8645)

**Solution:**
I found that it relates to this line of code in appendRowHTML for slick.grid.ts:

```
   const frozenRowOffset = this.getFrozenRowOffset(row);
    const rowDiv = Utils.createDomElement('div', { className: `ui-widget-content ${rowCss}`, role: 'row', style: { top: `${this.getRowTop(row) - frozenRowOffset}px` } });
    let rowDivR: HTMLElement | undefined;
    divArrayL.push(rowDiv);
    if (this.hasFrozenColumns()) {
      // it has to be a deep copy otherwise we will have issues with pass by reference in js since
      // attempting to add the same element to 2 different arrays will just move 1 item to the other array
      rowDivR = rowDiv.cloneNode(true) as HTMLElement;
      divArrayR.push(rowDivR);
    }

```

The following fix resolves the issue, and the repo tests still seem ok:

```
      const frozenRowOffset = this.getFrozenRowOffset(row);
      const topValue = this.getRowTop(row) - frozenRowOffset; // Calculated top value
      const rowDiv = Utils.createDomElement('div', {
        className: `ui-widget-content ${rowCss}`,
        role: 'row',
        style: { transform: `translateY(${topValue}px)` } // Using transform instead of top
      });
      
```

![image](https://github.com/user-attachments/assets/64c080a1-81d7-4d24-9965-6a0e8f74c7be)

If this all looks ok to you, wondering if a special access is required to push a branch with the fix ? As I tried to but it said access denied.

Thanks,
Pete

### Reproduction

Grid with approx 10m rows with grid line styles . - > close to 10m it will encounter rendering issues.

### Which Framework are you using?

React

### Environment Info

```shell
System:
    OS: Linux 6.8 Ubuntu 24.04 LTS 24.04 LTS (Noble Numbat)
    CPU: (24) x64 12th Gen Intel(R) Core(TM) i9-12900KF
    Memory: 51.57 GB / 62.58 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 20.15.1 - ~/.nvm/versions/node/v20.15.1/bin/node
    Yarn: 1.22.22 - ~/.nvm/versions/node/v20.15.1/bin/yarn
    npm: 10.8.2 - ~/.nvm/versions/node/v20.15.1/bin/npm
  Browsers:
    Chrome: 126.0.6478.182
```

### Validations

- [X] Follow our [Code of Conduct](https://github.com/6pac/SlickGrid/blob/master/CODE_OF_CONDUCT.md)
- [X] Read the [Wikis](https://github.com/6pac/SlickGrid/wiki).
- [X] Check that there isn't [already an issue](https://github.com/6pac/SlickGrid/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/6pac/SlickGrid/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ghiscoding** [comment]
This seems slightly related to issue #297

cc @6pac 

So in your code, you seem to completely replace the `top` style with `transform` of an Y value? What happens to the next few lines for the frozen column with the clone node? You're not showing them in the bottom portion of the code change, so I'm unsure if you're removing them or what not!?

With this change however, this will most probably break all Cypress E2E tests because a few of these tests rely on the `top` style to exist and query against it for the test, for example because the SlickGrid Virtual Scroll can replace and shift rows in the DOM (not visually but in the DOM tree), I rely on the `top` to get whatever rows I want. What I mean is that if you filter some values in SlickGrid and you scroll down and up, you can't rely on let say give me 3rd row of `querySelectorAll('.slick-row')` because this might give you row 10 instead, however we can use `top` to get 3rd row (row height * 3 => `top: 75px`) which is convenient in a Cypress test and I use it often... so changing `top` to `transform` might have a lot of implications



> wondering if a special access is required to push a branch with the fix ? As I tried to but it said access denied.

That's the default behavior of GitHub and any other repos using git, no one in the world has access to push directly to a repo. You need to fork the repo and then push to your fork. After pushing to your fork, it will then ask you if you want to create a Pull Request to push to the original repo (here). So it's not a 1 step push, but rather a 3 steps push (unless you're a contributor/maintainer of this repo which is only 6pac and I)

**Comment by ghiscoding** [comment]
Yes some people use it with extremely large dataset, I did see someone using that much data but I would never go with that much size myself (it's not recommended). 

I already mentioned exactly why the Cypress tests will fail, I'm often using the `top` with a `querySelector` to find the x row in the grid and now you want to completely remove `top` and replace it with `transform`, so yes of course that will fail the tests and potentially break a few user's code. This really looks like a Breaking Change to me

I don't want to use compare tools to find out what really change, it would be helpful if you could use the ` ```diff` instead and add `+` for new/change with lines of code and `-` for deleted/changed previous lines of code with, for example

```diff
- hello world
+ Hello World!
```

I'm not too sure that I want to proceed considering that this is most probably a Breaking Change.

**Comment by ghiscoding** [comment]
Yeah sure I'm open to comment and discussions, I just think that we have to consider the potential side effect in doing this kind of change because that could impact end users which I always try to avoid (even if it fixes a bug, it might be a Breaking Change). So I thought your change was just a few lines but it looks to be many more lines than I expected. 

So in the end, I didn't mean that I will refuse the PR, but I meant to say is that if it's really the only way to go, then that would require a `major` version v6.0 because of the Breaking Change (changing from `top` to `transform`) and that is because of the potential side effect that could break user's code or expectations. However doing a major version for this small change seems a little too much... so my question is, is it really the only way to fix this issue? isn't there a way that we can keep the `top` somehow and still fix the issue without requiring a major version? If it's really the only way, then I find the list of changes for a major version to be very light.

**Comment by 6pac** [comment]
From my point of view, the issues you raise are very valid and I'd definitely support working on them.  
It probably remains to be seen what the impact would be on the codebase, but if there were very tangible visual improvements, then I'd say it would be worth some degree of pain to implement them.

I'll let @ghiscoding speak for himself, but I suspect he'll say the same thing.

**Comment by ghiscoding** [comment]
There's a throttle in place which makes the virtual scroll wait before rendering the next batch. It's currently set to 50ms, you can try to decrease it. I think it was added by Ben a long time ago (it doesn't seem to exist in the original SlickGrid repo though). You have a large dataset so you are the best person who could give it a try with maybe 0 and see what happens (there's more info in old PR #810). 

https://github.com/6pac/SlickGrid/blob/87b8c6af4bed91bb646971b96d81185f346e7d2c/src/slick.grid.ts#L290

I actually always wondered why that was put in place, so perhaps you could try to disable it completely, something like this maybe. So if you comment out the following line (and where it's being used, which is the `enqueue` and `dequeue`). There's a good chance that it would make the rendering experience worst which would help explain why this code was put in place, but without really giving it a try on a large dataset like yours, it's hard to know for sure without testing it.

https://github.com/6pac/SlickGrid/blob/87b8c6af4bed91bb646971b96d81185f346e7d2c/src/slick.grid.ts#L663

https://github.com/6pac/SlickGrid/blob/87b8c6af4bed91bb646971b96d81185f346e7d2c/src/slick.grid.ts#L4918

https://github.com/6pac/SlickGrid/blob/87b8c6af4bed91bb646971b96d81185f346e7d2c/src/slick.grid.ts#L5105


there's also a grid option for how many rows to keep in the buffer when scrolling, which in other words mean how many row do we want to cache and make ready before the scroll happens. I'm not sure if it has any impacts or not but maybe try to increase it a bit? 

https://github.com/6pac/SlickGrid/blob/87b8c6af4bed91bb646971b96d81185f346e7d2c/src/slick.grid.ts#L283



> Note I should have mentioned if it’s a matter of porting tests I’m happy to check that out at the same time.

it would be part of it, but the question is more about "Do we really want to do a Breaking Change version with just this change? or can we do a bug fix that will not introduce a Breaking Change". Because in my head, I'm really consider the `top` to `transform` to be a Breaking Change and I'm trying hard to not push a new major yet again, but if it's really the only solution that works then yeah we'll go ahead with it.

**Comment by 6pac** [comment]
not aware of ever having added the throttle, but if you have evidence, then I suppose it was me! Way back when this was happening, there was much less standardisation between browsers so it could well have been a cross-browser compromise. I think in the modern context we would be much more confident in the stability of a more optimal solution.

**Comment by ghiscoding** [comment]
> not aware of ever having added the throttle, but if you have evidence, then I suppose it was me! Way back when this was happening, there was much less standardisation between browsers so it could well have been a cross-browser compromise. I think in the modern context we would be much more confident in the stability of a more optimal solution.

haha yeah it's not about pointing fingers, but more about the reason why it was put in place. That might have been copied from another fork too, there was a bunch of different that existed at that point in time too so... 

anyway, you can see similar discussions we had about this a while ago in this [comment](https://github.com/6pac/SlickGrid/issues/219#issuecomment-1630077429) and the commit where you added this throttling is 93f043fa

**Comment by 6pac** [comment]
So I did. I'm pretty sure I didn't write that code though - it must have been either patched in from another repo or contributed in an issue. So in summary, I don't know exactly why it was added. It does definitely make sense to have *some* throttling for long scrolls, but there appears to be some room for improvement.

**Comment by ghiscoding** [comment]
> anyway, you can see similar discussions we had about this a while ago in this [comment](https://github.com/6pac/SlickGrid/issues/219#issuecomment-1630077429) and the commit where you added this throttling is [93f043f](https://github.com/6pac/SlickGrid/commit/93f043fa587e377462d2fe0fa7525ff9f1b3f3ef)

@pbower  there's a bit more info in that comment (and thread) that I referenced, it seems like 5ms was a good number to use when I tested it at that time.

**Comment by ghiscoding** [comment]
ohhh wait a second, I just found how to push this into a new version without making a Breaking Change. We could simply add this into a grid option keep `top` as the default and that's it. You'll be able to change it to a `transform` while not breaking other expectations and everyone is happy.. woohoo! 🎉 something like this in the `GridOption` interface `topStyleRenderType: 'top' | 'transform'` (I'm not too crazy about the name, I'm sure you can find better) would be the default and then just use either or in the code...

```diff
 const rowDiv = Utils.createDomElement('div', {
      className: `ui-widget-content ${rowCss}`,
      role: 'row',
-      style: { transform: `translateY(${topValue}px)` } // Using transform instead of top
});

+ if (options.topStyleRenderType === 'transform') {
+   rowDiv.style.transform = `translateY(${topValue}px)`;
+ } else {
+   rowDiv.style.top = `${topValue}px`;
+ }
```
and voila! Perf improvements is now doable on your side and it won't break users who still expect `top` because that would remain the default. If we ever decide to do a `major` version in the future, if it ever happen, then we could switch the default to `transform`

**Comment by 6pac** [comment]
@pbower might be woth checking this out too: https://www.thecandidstartup.org/2024/04/29/modern-react-virtual-scroll-grid-9.html

**Comment by 6pac** [comment]
Yep. From what I can tell, former rockstar programmer gets tied of the stress, makes a well compensated exit and had a lot of time on his hands. My dream.

**Comment by ghiscoding** [comment]
@pbower we didn't hear from your side, so I went ahead and done couple of small PRs related to this issue
- added a new grid option `rowTopOffsetRenderType?: 'top' | 'transform'`, so that you can start using it without breaking other users since the default remains `top` (see PR #1050)
https://github.com/6pac/SlickGrid/blob/dc08ab801ccd69b4875bcaf22b2c36a80fc192b7/src/slick.grid.ts#L3950-L3957
- decreased `scrollRenderThrottling` grid option to 10ms instead of previously 50ms, this should reduce the time it shows a blank section when scrolling to some rows that are uncached (for example scroll completely to the bottom of the grid)
- I also found that `forceSyncScrolling` grid option (defaults to false) exists to completely bypass the `scrollRenderThrottling`

https://github.com/6pac/SlickGrid/blob/dc08ab801ccd69b4875bcaf22b2c36a80fc192b7/src/slick.grid.ts#L5113-L5118

This is all available in [v5.12.0](https://github.com/6pac/SlickGrid/releases/tag/5.12.0)

**Comment by ghiscoding** [comment]
@pbower Side note on the grid option that I added `rowTopOffsetRenderType?: 'top' | 'transform'`, it actually looks like Ag-Grid is also using `transform` for perf reason. I found that out lately when I added the recent `rowspan` feature, taking a look at Ag-Grid, they had this note on Row Span

> Ag-Grid - [Row Spanning](https://www.ag-grid.com/angular-data-grid/row-spanning/)
> Configuring Row Spanning 
> To allow row spanning, the grid must have property suppressRowTransform=true. Row spanning is then configured at the column definition level. To have a cell span more than one row, return how many rows to span in the callback colDef.rowSpan.

Their note is also true for our `rowspan` as well, it doesn't work well with `transform` and we need to use `top` for that feature, but for the rest using `transform` seems to have better perf as per what Ag-Grid is mentioning above.

So anyway, it's worth knowing about it 😃  
@pbower have you tried the new grid option I added yet `rowTopOffsetRenderType: 'transform'`?

**Comment by ghiscoding** [comment]
Nice to see you're making progress in improving the performance of the grid. I can't see the videos though, probably because I'm not logged in. The failing tests all seems to be for the Row Detail plugin, which is a bit special, it pushes all the next rows further down when the Row Detail is expanded (open), so this might need adjustment with the cache that you mentioned.

**Comment by ghiscoding** [comment]
> Here's a gif snippet, that funnily enough accentuates the difference due to the lower frame rate.

that does seem a lot better on the UI side, wow nice job. I assume your improved code is on the left side... correct? Also curious, how large is your dataset? is that 10 million rows like you mentioned when you've opened the issue?

**Comment by ghiscoding** [comment]
Some side notes to be aware of...

> **Minimising Direct DOM Manipulations** – Instead of updating the DOM element by element, changes are batched and inserted in a single operation (using techniques like DocumentFragment). This reduces layout reflows and repaints.

not all systems/browsers/frameworks support DocumentFragment (for example it's not supported in Salesforce environment, which is super weird and is our biggest usage), so to counter that I added this grid option `preventDocumentFragmentUsage` that will use regular HTMLElement when the option is enabled. So it's something to consider



> **Caching and Reusing Computed Values** – Frequently used values (such as frozen row offsets, column positions, and computed CSS classes) are calculated once and reused, eliminating redundant calculations.

that's mainly what I've done to implement the recent `rowspan` feature, I added an object that keeps a cache of all spans. I copied that technique for the row cache that is built into SlickGrid as well. Then at some point or use case these caches are updated/recreated, I'm guessing you have something similar here...



> **Caching and Reusing Computed Values** – Frequently used values (such as frozen row offsets, column positions, and computed CSS classes) are calculated once and reused, eliminating redundant calculations.
> 
> **Efficient Cloning** – In cases where a row needs to be rendered into multiple canvases (e.g., frozen columns), the new approach clones existing nodes instead of recreating them from scratch. This reduces processing overhead and ensures consistency.
> 
> **Batching Render Operations** – Rows and cells are collected in arrays and only appended in batches, reducing the number of individual DOM insertions. This is particularly beneficial when rendering a large number of elements.

how much code size is that adding to the library though? I mean it shouldn't be a concern since we're gaining a boost in perf but it's still worth knowing since we obviously don't want to double its size. 


For the failing Row Detail failures, it's probably a matter of updating the Row Detail plugin which like I said pushes x amount of rows further down (or up if you're collapsing), so we might need some kind of different ways to invalidate this new cache that you're talking about? Without looking at the code, I'm assuming the Row Detail might be calling `invalidateRow(x)` for each row impacted by the open/collapse of a Row Detail, or a full `invalidate()` which would re-render all visible rows... So anyway, we probably need something to invalidate the new cache (ideally the existing invalidate functions should automatically do that which is what I've tried to do as much as possible with the new `rowspan`, in some cases it's cheaper to do a full `invalidate()` and rebuild the cache since it's only the visible rows).

**Comment by ghiscoding** [comment]
@pbower do you have any updates on the change you started? If it's only problems related to Row Details, I could certainly take a look. It might be interesting if you could open a new PR as a Draft that could get us started with testing, investigating and troubleshooting. Cheers

**Comment by ghiscoding** [comment]
I added an optional formatted data caching (cache formatter parsing results) in my own libs Slickgrid-Universal (and other frameworks) and that helped tremendously especially in Excel Export and also in UI when using lots of columns. See https://github.com/ghiscoding/slickgrid-universal/pull/2563 for more info, I won't implement it here though
